### PR TITLE
Add bundle lock --update --bundler

### PR DESCRIPTION
**Pull request by alyssais**
### What was the end-user problem that led to this PR?

The problem was that I couldn't generate a Bundler 2.0 lockfile on a machine that doesn't have the libraries available to be able to build native extensions, which is part of my normal workflow.

### What was your diagnosis of the problem?

My diagnosis was that `bundle lock --update`, which I can usually use instead of `bundle update`, didn't support the `--bundler` switch like `bundle update` does.

### What is your fix for the problem, implemented in this PR?

My fix implements the `--bundler` flag for `bundle lock --update`.

### Why did you choose this fix out of the possible options?

I chose this fix because I think it makes sense for `bundle lock --update` to remain consistent with `bundle update`, except for not actually installing gems.

### CODE REVIEW

**Comment by alyssais** [comment]
For this to be useful, RubyGems will also need a patch:

```diff
diff --git a/lib/rubygems/bundler_version_finder.rb b/lib/rubygems/bundler_version_finder.rb
index e74baca1..e7148d3f 100644
--- a/lib/rubygems/bundler_version_finder.rb
+++ b/lib/rubygems/bundler_version_finder.rb
@@ -49,7 +49,8 @@ To install the missing version, run `gem install bundler:#{vr.first}`

   def self.bundle_update_bundler_version
     return unless File.basename($0) == "bundle".freeze
-    return unless "update".start_with?(ARGV.first || " ")
+    return if ARGV.first.nil?
+    return if %w[update lock].none? { |c| c.start_with?(ARGV.first) }
     bundler_version = nil
     update_index = nil
     ARGV.each_with_index do |a, i|
```

Makes sense to do it here first though, I think.

File: spec/commands/lock_spec.rb [hunk 1]
```diff
@@ -360,3 +360,24 @@ def read_lockfile(file = "Gemfile.lock")
...
+
+    install_gemfile! <<-G
+      source "file:#{gem_repo4}"
+      gem "rack"
+    G
+    lockfile lockfile.sub(/(^\s*)#{Bundler::VERSION}($)/, '\11.0.0\2')
+
+    FileUtils.rm_r gem_repo4
+
+    bundle! :update, :bundler => true, :verbose => true
```
Comment (deivid-rodriguez) [inline spec/commands/lock_spec.rb]: Should be actually testing `bundle lock --update --bundler` here? Maybe we can add a test similar to 

https://github.com/bundler/bundler/blob/d0103bf094fbde0f38955862872c245e847612d4/spec/commands/lock_spec.rb#L78-L85

but for `bundle lock --update --bundler` instead?

**Review by deivid-rodriguez — CHANGES_REQUESTED** [review]

**Comment by deivid-rodriguez** [comment]
Sorry for the delay @alyssais.

So... if `bundle lock --update` always updates the bundler version, why do we need this new flag? I didn't fully understand the usefulness in the "bundler version finder" side. I think the only reason for the patch you posted in https://github.com/bundler/bundler/pull/6925#issuecomment-457156864 would be to skip usages of this new option when selecting the bundler version to use, but if we don't need the new option, we don't need to patch the bundler version finder either... right?

**Comment by alyssais** [comment]
So, currently, if I have Bundler 1.x and 2.x installed, and I want to update a lockfile from 1.x to 2.x, my only option is to run `bundle update --bundler`. If I run anything else, RubyGems' `BundlerVersionFinder` will intercept, and run Bundler 1.x, so Bundler won't be updated. There is currently no way to update the Bundler version of a lockfile without installing all the gems.

I should be able to run _something_ that has the same Bundler-updating effect, without installing all the gems in my Gemfile. Normally the command for this would be `bundle lock --update`, but because of `BundlerVersionFinder`, that will run Bundler 1.x and not update Bundler.

So, `BundlerVersionFinder` needs to be modified in some way to allow some command that can handle this case, hence the patch.

Now, initially I had planned to add `bundle lock --update --bundler` for consistency with `bundle update --bundler`, but it seems that, unlike `bundle update`, `bundle lock --update` _always_ updates Bundler anyway. So, there's no need for `--bundler`. There is still a need for `BundlerVersionFinder` to be modified to allow running `bundle lock` as well as `bundler update`, though.

So, _I think_, what should happen here is I should close this PR, and just send the `BundlerVersionFinder` patch to RubyGems. Does that sound right?

It is, however, extremely possible I've got something wrong here – I find the implementation of autoswitching across Bundler and RubyGems quite difficult to keep track of in my head, especially given it's been a month since I last had to think about it, and several months since my initial work on this.

**Comment by deivid-rodriguez** [comment]
@alyssais That makes sense to me, I understand now :+1:.

What the bundler version finder in rubygems seems to be doing is autoswitching on `bundle update`_unless_ you're running `bundle update --bundler`. So I think it makes sense to add the flag here as proposed initially, that only updates the bundler version in the lock file, and explicitly guard for that case in the bundler version finder.

I plan to review some issues with the autoswitching like the one you mentioned about `bundle lock --update` updating the locked bundler version, but I think this PR is good to go once we get it properly tested. I think it's better to keep this consistent, and we can keep the autoswitching in rubygems as it's working now for `bundle update`, without excluding all `bundle lock` variants from it.

> especially given it's been a month since I last had to think about it, and several months since my initial work on this.

I'm so sorry I've been so slow getting back to you here... :pray:.

**Comment by alyssais** [comment]
> I plan to review some issues with the autoswitching like the one you mentioned about `bundle lock --update` updating the locked bundler version, but I think this PR is good to go once we get it properly tested.

Given the `bundle lock --update` behaviour you mentioned, I don't think it will be possible to write a test for `bundle lock --update --bundler` that doesn't also pass for `bundle lock --update` – I'm guessing this is okay for now?

> I'm so sorry I've been so slow getting back to you here... 🙏.

No worries! Wasn't trying to criticise (I've been slow too), just to explain why my replies are so confused!

**Comment by deivid-rodriguez** [comment]
> Given the bundle lock --update behaviour you mentioned, I don't think it will be possible to write a test for bundle lock --update --bundler that doesn't also pass for bundle lock --update – I'm guessing this is okay for now?

I would expect the new flag to _only_ update the bundler version. If that's the idea, it should be possible to write a test that includes some other potential updates that are skipped when running `bundle lock --update --bundler`?

**Comment by alyssais** [comment]
> I would expect the new flag to _only_ update the bundler version.

I think I would expect `bundle lock --bundler` to do this, but `bundle lock --update --bundler` to behave as `bundle lock --update` does now.

The reason being that `bundle update --bundler` updates all your gems in addition to Bundler, so `bundle lock --update --bundler` should be consistent with that.

**Comment by deivid-rodriguez** [comment]
That makes sense, yeah :+1:. No problem if `bundle lock --update --bundler`, and `bundle lock --update` can't currently be distinguished in the tests.

**Comment by alyssais** [comment]
Cool. I'll see what I can do. Thanks for bearing with me!

**Comment by deivid-rodriguez** [comment]
Hello again @alyssais! Did you find time to work on this?

**Pull request closed without merging**